### PR TITLE
Flow analysis: specify for loops and fix types of interest for for/while loops

### PR DESCRIPTION
**Pull request by stereotype441**
Fixes https://github.com/dart-lang/language/issues/1203.  Corresponding implementation change is here: https://dart-review.googlesource.com/c/sdk/+/162483

### CODE REVIEW

File: resources/type-system/flow-analysis.md [hunk 1]
```diff
@@ -683,7 +705,15 @@ TODO: Add missing expressions, handle cascades and left-hand sides accurately
   (E) S` then:
   - Let `before(E) = conservativeJoin(before(N), assignedIn(N), capturedIn(N))`.
   - Let `before(S) = split(true(E))`.
-  - Let `after(N) = join(false(E), unsplit(break(S))`
+  - Let `after(N) = inheritTested(join(false(E), unsplit(break(S))), after(S))`.
+
+- **for statement**: If `N` is a for statement of the form `for (D; C; U) S`,
+  then:
+  - Let `before(D) = before(N)`.
+  - Let `before(E1) = conservativeJoin(after(D), assignedIn(N), capturedIn(N))`.
```
Comment (eernstg) [inline resources/type-system/flow-analysis.md:713]: Wouldn't that be 'Let `before(C) = ...`'?

File: resources/type-system/flow-analysis.md [hunk 2]
```diff
@@ -683,7 +705,15 @@ TODO: Add missing expressions, handle cascades and left-hand sides accurately
   (E) S` then:
   - Let `before(E) = conservativeJoin(before(N), assignedIn(N), capturedIn(N))`.
   - Let `before(S) = split(true(E))`.
-  - Let `after(N) = join(false(E), unsplit(break(S))`
+  - Let `after(N) = inheritTested(join(false(E), unsplit(break(S))), after(S))`.
+
+- **for statement**: If `N` is a for statement of the form `for (D; C; U) S`,
```
Comment (eernstg) [inline resources/type-system/flow-analysis.md:710]: Woohoo! ;-)

File: resources/type-system/flow-analysis.md [hunk 3]
```diff
@@ -683,7 +705,15 @@ TODO: Add missing expressions, handle cascades and left-hand sides accurately
...
+  then:
+  - Let `before(D) = before(N)`.
+  - Let `before(E1) = conservativeJoin(after(D), assignedIn(N), capturedIn(N))`.
+  - Let `before(S) = split(true(C))`.
+  - Let `before(U) = merge(after(S), continue(S))`.
+  - Let `after(N) = inheritTested(join(false(E), unsplit(break(S))), after(U))`.
```
Comment (eernstg) [inline resources/type-system/flow-analysis.md:716]: Wouldn't `E` be `C`?

Also, can we now add a new bullet point along the lines of this?:

```
- **for-in statement**: If `N` is a for-in statement of the form
  `for (D in I) S` then the flow analysis is performed according
  to the desugaring of `N`. *The desugaring is specified in the
  language specification, and it yields a `while` loop.*
```

**Review by eernstg — APPROVED** [review]
LGTM, with the suggestion that we might add a bullet point about for-in statements, and pointing out what seems to be a couple of typos.

**Comment by eernstg** [comment]
Removed the comment about for-in statements, it's being handled in #1219.

File: resources/type-system/flow-analysis.md (hunk 1 above)
Comment (stereotype441) [reply resources/type-system/flow-analysis.md]: Fixed, thanks!

File: resources/type-system/flow-analysis.md (hunk 3 above)
Comment (stereotype441) [reply resources/type-system/flow-analysis.md]: Fixed!

**Pull request merged**